### PR TITLE
[4.3] Message after a login / logout frontend

### PR DESCRIPTION
**Pull request by sandewt**
Pull Request for Issue # .

New feature, see #30767

### Summary of Changes

There is a message after a successfully login and logout in the frontend.
### Messages
new_feature_messages_3
new_feature_messages_2
[EDIT] Removed _successfully_ from the text in message.

### Users: Options
new_feature_messages_4
[EDIT] Removed

### Testing Instructions

Backend: Go to Users > Manage > Options
(Frontend: Is not applicable.)

Set the options:
- Frontend Login Message (= Yes, default = No)
- Frontend Logout Message (= Yes, default = No)

Create a menu link:
- Menu Login form
- Menu Logout form

1) Log in and out, in case of a:
- Login form (default)
- Menu Login form
- Menu Logout form

2) Code review (developers);

- Check if the code is correct and secure.

### Actual result BEFORE applying this Pull Request

There is NOT showing a message after a suscessfully login or logout.

### Expected result AFTER applying this Pull Request

- There is showing a message after a suscessfully login. 
- There is showing a message after a suscessfully logout.

### Documentation Changes Required

Yes.

### CODE REVIEW

**Comment by HLeithner** [comment]
Please rebase on 4.1-dev

File: administrator/components/com_users/config.xml [hunk 1]
```diff
@@ -122,6 +122,30 @@
 			<option value="2">COM_USERS_CONFIG_FIELD_ENFORCE_2FA_FIELD_ADMIN</option>
 			<option value="3">COM_USERS_CONFIG_FIELD_ENFORCE_2FA_FIELD_BOTH</option>
 		</field>
+		
+		<field
+			name="frontend_login"
```
Comment (SharkyKZ) [inline administrator/components/com_users/config.xml]: Can you change this to something less misleading? E.g. `frontend_login_message`.

Comment (sandewt) [reply administrator/components/com_users/config.xml]: Done

**Comment by sandewt** [comment]
>Please rebase on 4.1-dev

It's just a small extension, a new feature might be a bit exaggerated. Or is there some other argument?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

File: language/en-GB/com_users.ini [hunk 2]
```diff
@@ -31,6 +31,8 @@ COM_USERS_FIELD_RESET_CONFIRM_USERNAME_LABEL="Username"
 COM_USERS_FIELD_RESET_PASSWORD1_LABEL="Password"
 COM_USERS_FIELD_RESET_PASSWORD1_MESSAGE="The passwords you entered do not match. Please enter your desired password in the password field and confirm your entry by entering it in the confirm password field."
 COM_USERS_FIELD_RESET_PASSWORD2_LABEL="Confirm Password"
+COM_USERS_FRONTEND_LOGIN_SUCCESS="You have been successfully logged in."
```
Comment (Quy) [inline language/en-GB/com_users.ini]: Remove `successfully`. See https://github.com/joomla/joomla-cms/pull/14607

**Comment by maikol-ortigueira** [comment]
I have tested this item :white_check_mark: successfully on 963ad7deca4ef06a722b95009dac832fa42b7a98<br /><br />I have tested the patch and it works perfectly<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by pabloarias** [comment]
I have tested this item :white_check_mark: successfully on 963ad7deca4ef06a722b95009dac832fa42b7a98<br /><br />Tested successfully on Joomla 4.0.0-beta5-dev and PHP 7.3.22-1

Thank you!

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

File: language/en-GB/com_users.ini (hunk 2 above)
Comment (sandewt) [reply language/en-GB/com_users.ini]: Done.

**Comment by alikon** [comment]
i'll not set this PR as RTC despite  2 good tests cause of https://github.com/joomla/joomla-cms/pull/30834#issuecomment-701965397
i'll let the final decision to RL's

**Comment by brianteeman** [comment]
We might as well all go home if this wont get merged

**Comment by sandewt** [comment]
>Some checks were not successful: continuous-integration/drone/pr — Build is failing

What does this mean? Action?

[EDIT] I see that this problem has resolved itself.

**Comment by HLeithner** [comment]
> We might as well all go home if this wont get merged

We have 27 release blocker and most of the time I only see some new stuff if no one fixes the release blocker and only add new fancy stuff then you are right and we can go home.
https://github.com/joomla/joomla-cms/issues?q=is%3Aopen+is%3Aissue+label%3A"Release+Blocker"

**Comment by sandewt** [comment]
> We have 27 release blocker and most of the time I only see some new stuff if no one fixes the release blocker and only add new fancy stuff then you are right and we can go home.

I will close this PR and make a new PR based on [4.1].

**Comment by HLeithner** [comment]
You don't need to, I changed your target branch to 4.1-dev

**Comment by sandewt** [comment]
> You don't need to, I changed your target branch to 4.1-dev

I see it has already happened. :smiley:

Thanks.

**Comment by sandewt** [comment]
>Please rebase on 4.1-dev

Please test in Joomla [4.1].

@maikol-ortigueira / @pabloarias<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by tushar33** [comment]
I have tested this item :white_check_mark: successfully on 91ccc4f8401bbb4c22c0c07caae855812d173839<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by particthistle** [comment]
I have tested this item :white_check_mark: successfully on 91ccc4f8401bbb4c22c0c07caae855812d173839<br /><br />Nice addition @sandewt. Too many times have I not realised I was logged in or out when wanting to be the reverse after clicking the button in J3.

Testing scenarios all worked.

Please add Documentation Required tag so that it can be noted to go back and review relevant help and JDocs.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Review by zero-24 — APPROVED** [review]
LTGM @bembelimen please take a look here too :+1:

**Comment by bembelimen** [comment]
I really like the feature (good idea) but I'm not sure about the parameter tbh. Is it needed at all? Perhaps we should (in a future version) go for a message management extension where all this notifications can be managed on one point?
Any thoughts?

**Comment by zero-24** [comment]
What kind of `message management extension` are you looking for? Shouldt a module / plugin / component take care of its own messages?

**Comment by brianteeman** [comment]
Something like the mail templates?

**Comment by bembelimen** [comment]
Yes exactly.

**Comment by snehaM26** [comment]
I have tested this item :white_check_mark: successfully on 91ccc4f8401bbb4c22c0c07caae855812d173839<br /><br />I have tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by ashvini77** [comment]
I have tested this item :white_check_mark: successfully on 91ccc4f8401bbb4c22c0c07caae855812d173839<br /><br />Its showing a message on login and logout on frontend.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by sandewt** [comment]
Thanks all.

**Comment by bembelimen** [comment]
Again: I would merge it, but I think the parameter are not necessary. We should avoid to add parameter for each thing (yeah I know we did it in the past very often...). If someone want to disable the messages, he/she/it can just make a language override to an empty string.

Additionally I don't think it's necessary to check if user is really logged in, otherwise the login/logout would have returned before, so you can assume, they are.

Could you adjust the PR for this?

**Comment by sandewt** [comment]
>Again: I would merge it, I would merge it, but I think the parameter are not necessary.

Sorry, I read about this. Do you mean both parameters?
For now I would like to leave it that way, until there is a centrally regulated message system. Or if several want it that way.

>Additionally I don't think it's necessary to check if user is really logged in, otherwise the login/logout would have returned before, so you can assume, they are.

You mean?

Change this code:
```
// Show a message when a user is logged in.
if ($login === 1 && Factory::getUser()->get('id') > 0)
```
to:
```
// Show a message when a user is logged in.
if (ComponentHelper::getParams('com_users')->get('frontend_login_message', 0))
{
```

**Comment by bembelimen** [comment]
> Do you mean both parameters?

Yes both parameters.

**Comment by sandewt** [comment]
>If someone want to disable the messages, he/she/it can just make a language override to an empty string.

I have the following notes:

1) Not everyone likes a login / outlog message and can get irritated by it. Hence the default parameters are hide.
2) Especially novice users of Joomla are often unfamiliar with overrides.

**Comment by sandewt** [comment]
> Additionally I don't think it's necessary to check if user is really logged in, otherwise the login/logout would have returned before, so you can assume, they are.

Done. Deleted check if user is logged in<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by Tejashrimajage** [comment]
I have tested this item :white_check_mark: successfully on ce1bf617029d4b3ffd244a0f223d74ec78c1f78e<br /><br />It's working fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by vaibhavsTekdi** [comment]
I have tested this item :white_check_mark: successfully on ce1bf617029d4b3ffd244a0f223d74ec78c1f78e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by HLeithner** [comment]
@sandewt please remove the parameters they are unneeded and using the override component should be possible with documentation thx

**Comment by richard67** [comment]
Needs some change as stated above.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by sandewt** [comment]
> Removed: ? ? ? ? 

See [https://github.com/joomla/joomla-cms/pull/30834#issuecomment-968072375](https://github.com/joomla/joomla-cms/pull/30834#issuecomment-968072375 "https://github.com/joomla/joomla-cms/pull/30834#issuecomment-968072375")<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by richard67** [comment]
Back to pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by sandewt** [comment]
> Back to pending.

@HLeithner and @bembelimen 

Deleted the two parameters.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by sandewt** [comment]
Please test.

- Install the patch.
- Login and logout and see the messages.

That's all, the options have been removed.

**Comment by Quiviro** [comment]
I have tested this item :white_check_mark: successfully on 61634a3cd83883e5204f67f8a98f0cf8bf4480da<br /><br />The messages are shown after login/logout<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by BertaOctech** [comment]
I have tested this item :red_circle: unsuccessfully on 61634a3cd83883e5204f67f8a98f0cf8bf4480da<br /><br />I am sorry but after applyingg the patch I do not see the messages

I am testing on 4.1.0-rc1

Some system info:
------------
 Database Type 	mysql
Database Version 	10.4.19-MariaDB
Database Collation 	utf8mb4_general_ci
Database Connection Collation 	utf8mb4_general_ci
Database Connection Encryption 	None
Database Server Supports Connection Encryption 	No
PHP Version 	7.4.20
Web Server 	Apache/2.4.48 (Unix) OpenSSL/1.1.1k PHP/7.4.20 mod_perl/2.0.11 Perl/v5.32.1
WebServer to PHP Interface 	apache2handler
Joomla! Version 	Joomla! 4.1.0-rc1 Release Candidate [ Kuamini ] 18-January-2022 18:00 GMT
User Agent 	Mozilla/5.0 (X11; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0 
---------------<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by sandewt** [comment]
>I am sorry but after applyingg the patch I do not see the messages
>I am testing on 4.1.0-rc1

I have no explanation for this.

Can you test with Joomla! **4.1.0-dev** Development [ Kuamini ] ?

[EDIT] I have tested successfully on:  Joomla! **4.1.0-rc1** Release Candidate [ Kuamini ] 18-January-2022 18:00 GMT

**Comment by BertaOctech** [comment]
Hi again.

I am sorry but could not test the patch successfully.

I took a fresh install from https://developer.joomla.org/nightly-builds.html
I configured the patch tester but for some reason the oldest pull id available to me was #36134

I then tried to apply the patch manually but I did not get the expected result.

Best regards,

Berta

**Comment by pabloarias** [comment]
I have tested this item :white_check_mark: successfully on 61634a3cd83883e5204f67f8a98f0cf8bf4480da<br /><br />Hello:

My currently installed Joomla! version is ‎4.1.0-rc3-dev. I'm using PHP 8.0.15.

I've created a menu item to login and another to logout.

Before apply path: it doesn't show messages after login or logout.

After applying the path: it shows a message in both cases. After login and logout. Via module or menu items.

Thank you!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 61634a3cd83883e5204f67f8a98f0cf8bf4480da<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by infograf768** [comment]
rtc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by sandewt** [comment]
Thanks all. :smiley:

**Comment by sandewt** [comment]
Oops, I see that this feature is not yet added to J4.1.
So, I have merged this branche.

**Comment by sandewt** [comment]
@richard67 

Why is this PR not merged to the core yet?

**Comment by richard67** [comment]
@sandewt No idea. I am not the only maintainer.

**Comment by laoneo** [comment]
This change should go into 4.2. I changed the base branch, but now there are some commits added to your pr which will disappear when 4.1 got fully merged up into 4.2.

**Comment by richard67** [comment]
> > Additionally I don't think it's necessary to check if user is really logged in, otherwise the login/logout would have returned before, so you can assume, they are.
> 
> Done. Deleted check if user is logged in

@sandewt I see the check has been deleted in the login method but not in the logout method. Does that have a reason, or was that just missed?

**Comment by sandewt** [comment]
@richard67 A good comment.
I don't know if there is a check somewhere in Joomla if the user is really logged out at that time. I'll go for safety then.

**Comment by bembelimen** [comment]
> @richard67 A good comment. I don't know if there is a check somewhere in Joomla if the user is really logged out at that time. I'll go for safety then.

No need to be "safe" here, as if the logout failed, it would have been caught before. Although the check would be: ```if ($app->getIdentity()->guest)``` (we shouldn't introduce deprecated code) but as mentioned it's not needed.

File: components/com_users/src/Controller/UserController.php [hunk 3]
```diff
@@ -228,6 +232,12 @@ public function logout()
 			$return = Uri::root();
 		}
 
+		// Show a message when a user is logged out.
+		if (Factory::getUser()->get('id') === 0)
```
Comment (laoneo) [inline components/com_users/src/Controller/UserController.php]:
```suggestion
		if ($app->getIdentity()->guest)
```

Comment (laoneo) [inline components/com_users/src/Controller/UserController.php:236]:
```suggestion
		if ($this->app->getIdentity()->guest)
```

**Comment by richard67** [comment]
@bembelimen @laoneo Do you think we need new tests after the last 2 commits?

**Comment by laoneo** [comment]
I think one test for confirmation is enough

**Comment by sandewt** [comment]
>I think one test for confirmation is enough

@laoneo I think not
```
echo Factory::getUser()->get('guest') . '<br>'; // let's see 1
echo Factory::getUser()->get('id') . '<br>'; // let's see  0
echo $app->getIdentity()->guest  . '<br>'; // let's see 0
echo $app->getIdentity()->id; // let's NOT see 0
```

**Comment by laoneo** [comment]
Are you also talking about the human tests for this pr?

**Comment by richard67** [comment]
I was indeed only talking about human tests.

**Comment by sandewt** [comment]
I think this PR is not successful for the above reasons. Please test.

**Comment by sandewt** [comment]
In summary:

When using this method, the user IS logged out (there is a logout message).

`if (Factory::getUser()->get('guest'))`

And when using the following method, the user is NOT logged out (there is no logout message).

`if ($app->getIdentity()->guest)`

Is completely unclear to me.

**Comment by bembelimen** [comment]
Just make it easy: remove the if... messages are enough

**Comment by sandewt** [comment]
>Just make it easy: remove the if... messages are enough

The moment the message 'You have been logged out.' is generated, the user is actually still logged in.
This is equivalent to flying the flag before finishing.
Of course, somewhere in the logout process, things will be checked, but still.

That's why I'm hesitating to drop the if statement. Give me a good reason to remove it.

**Comment by roland-d** [comment]
@sandewt 
> The moment the message 'You have been logged out.' is generated, the user is actually still logged in.

That is not the case right, the logout is called before that on line 167
https://github.com/joomla/joomla-cms/pull/30834/files#diff-d3df0595ae1fa944f7270d2b86a20ae9358c93d7206c2191da23aebd8ce585a0R167

Do you want to figure out the discrepancies between getting the guest status?

Your comment here https://github.com/joomla/joomla-cms/pull/30834#issuecomment-1067956161 may need some further investigation in a separate issue.

**Comment by richard67** [comment]
Back to pending (RTC was added back by the bot by accident).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30834">issues.joomla.org/tracker/joomla-cms/30834</a>.</sub>

**Comment by drmenzelit** [comment]
Thanks

**Comment by sandewt** [comment]
> Your comment here https://github.com/joomla/joomla-cms/pull/30834#issuecomment-1067956161 may need some further investigation in a separate issue.

@drmenzelit 

**Something went wrong here. This PR is merged to joomla 4.3. That shouldn't have happened.
Because this PR is not working.**

**Comment by sandewt** [comment]
Removing the if statement makes this PR work. The question is whether that is desirable. See the discussion above.

```
// Show a message when a user is logged out.
if ($app->getIdentity()->guest) {
    $app->enqueueMessage(Text::_('COM_USERS_FRONTEND_LOGOUT_SUCCESS'), 'message');
}
```

```
// Show a message when a user is logged out.
$app->enqueueMessage(Text::_('COM_USERS_FRONTEND_LOGOUT_SUCCESS'), 'message');
```

**Comment by bembelimen** [comment]
Please test #39048

**Comment by richard67** [comment]
@sandewt Please test #39048 . Thanks in advance.

**Comment by sandewt** [comment]
@richard67 

I'd like to do that, but then I'm actually testing my own code proposal. I wonder if that's desirable :wink:

**Pull request merged**